### PR TITLE
Add manifest format for bulk file downloads (#1931)

### PR DESCRIPTION
**Pull request by danielsotirhos**
Author (before primary review)

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue
- [x] Added `HCA` label <sub>or this PR does not target an `hca/*` branch</sub>
- [x] Created issue to track porting these changes <sub>or this PR does not need porting</sub> 
- [x] Added `[r]` prefix at start of commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label <sub>or this PR does not require reindexing</sub>
- [x] Freebies are blocked on this PR <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Ran `make requirements_update` <sub>or this PR does not change requirements*.txt</sub>
- [x] Mentioned `make requirements` in UPGRADING.rst <sub>or this PR does not change requirements*.txt</sub>
- [x] Documented upgrade of personal deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>

Primary reviewer (before pushing merge commit)

- [x] Checked `reindex` label and `[r]` prefix of commit title
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Build passes in `sandbox` <sub>or commented that sandbox will be skipped</sub>
- [x] Reindexed `sandbox` <sub>or this PR does not require reindexing</sub>
- [x] Added PR reference to merge commit
- [x] Moved linked issue to Merged
- [ ] Pushed merge commit to Github

Primary reviewer (after pushing merge commit)

- [ ] Moved freebies to Merged column <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened chain <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Commented on demo expectations <sub>or labeled as `no demo`</sub>
- [ ] Pushed merge commit to Gitlab
- [ ] Reindexed `dev` <sub>or this PR does not require reindexing</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Unassign reviewer from PR

### CODE REVIEW

**Comment by danielsotirhos** [comment]
```
curl -L "https://service.daniel.dev.singlecell.gi.ucsc.edu/manifest/files?catalog=dcp2&filters=%7B%22fileFormat%22%3A%20%7B%22is%22%3A%20%5B%22pdf%22%5D%7D%7D&format=bulk.download" | curl -L -K -
```

File: test/service/test_manifest.py [hunk 1]
```diff
@@ -1188,8 +1189,32 @@ def test_full_metadata_missing_fields(self):
...
+        self._index_canned_bundle(bundle_fqid)
+        filters = {'fileFormat': {'is': ['pdf']}}
+        response = self._get_manifest(ManifestFormat.bulk_download, filters)
+        self.assertEqual(200, response.status_code, 'Unable to download manifest')
+        lines = response.content.decode('utf-8').splitlines()
+        self.assertIn('--create-dirs', lines)
+        self.assertIn('--compressed', lines)
+        file_uuids, file_names = [], []
+        for line in lines:
+            if line.startswith('url='):  # url="https://.../repo/files/74897eb7-0701-4e4f-9e6b-8b9521b2816b"
```
Comment (achave11) [inline test/service/test_manifest.py:1205]: The comments in this for loop are a tad obscure, but maybe only for me. After reading the section thoroughly, I understand that the comments show what is expected.

**Review by achave11 — APPROVED** [review]
Looks good, I just had a comment but you may chose to disregard it. Great work!

File: test/service/test_manifest.py (hunk 1 above)
Comment (danielsotirhos) [reply test/service/test_manifest.py]: Good point. The comments were scaffolding used in the creation of the test and should really be removed. Thanks.

**Comment by danielsotirhos** [comment]
Taking PR out of review to rebase on latest `issues/danielsotirhos/1764-add-tdr-file-endpoint`

File: src/azul/service/manifest_service.py [hunk 2]
```diff
@@ -103,6 +103,7 @@ class ManifestFormat(Enum):
     compact = 'compact'
     full = 'full'
     terra_bdbag = 'terra.bdbag'
+    bulk_download = 'bulk.download'
```
Comment (hannes-ucsc) [inline src/azul/service/manifest_service.py:106]:
```suggestion
    curl = 'curl'
```

Likewise s/BulkDownloadManifest/CurlManifest/g.

File: src/azul/service/manifest_service.py [hunk 3]
```diff
@@ -597,6 +600,35 @@ def create_file(self) -> Tuple[str, Optional[str]]:
...
 
+class BulkDownloadManifestGenerator(StreamingManifestGenerator):
+
+    @property
+    def content_type(self) -> str:
+        return 'text/plain'
+
+    @property
+    def file_name_extension(self):
+        return 'txt'
```
Comment (hannes-ucsc) [inline src/azul/service/manifest_service.py:611]:
```suggestion
        return 'curlrc'
```

File: src/azul/service/manifest_service.py [hunk 4]
```diff
@@ -597,6 +600,35 @@ def create_file(self) -> Tuple[str, Optional[str]]:
...
+    @property
+    def entity_type(self) -> str:
+        return 'files'
+
+    def write_to(self, output: IO[str]) -> Optional[str]:
+        output.write('--create-dirs\n\n')
+        output.write('--compressed\n\n')
+        for hit in self._create_request().scan():
+            doc = self._hit_to_doc(hit)
+            assert isinstance(doc, dict)
```
Comment (hannes-ucsc) [inline src/azul/service/manifest_service.py:622]: Why is this necessary? The next line uses `[]` so it would fail if doc were not a dict.

File: test/service/test_manifest.py (hunk 1 above)
Comment (hannes-ucsc) [inline test/service/test_manifest.py]: This is too much post-processing of the response. This is very simple format and the index content is known. This can be a much more literal assertion like some of the other tests in this class.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: lambdas/service/app.py [hunk 5]
```diff
@@ -1319,6 +1319,10 @@ def get_order():
                   Participants (aka Donors) and one for Samples (aka Specimens).
                   For more on the format of the manifests see
                   [documentation here](https://software.broadinstitute.org/firecloud/documentation/article?id=10954).
+
+                - `curl` for a [curl configuration file](https://curl.haxx.se/docs/manpage.html#-K)
+                  manifest. This manifest can be used with the curl application
```
Comment (hannes-ucsc) [inline lambdas/service/app.py:1324]:
```suggestion
                  manifest. This manifest can be used with the `curl` program
```

File: test/service/test_manifest.py [hunk 6]
```diff
@@ -1195,6 +1195,28 @@ def test_full_metadata_missing_fields(self):
...
         self.assertGreater(len(symmetric_diff), 0)
 
+    @manifest_test
+    def test_curl_format_manifest(self):
+        self.maxDiff = None
+        bundle_fqid = BundleFQID('f79257a7-dfc6-46d6-ae00-ba4b25313c10', '2018-09-14T133314.453337Z')
+        self._index_canned_bundle(bundle_fqid)
+        filters = {'fileFormat': {'is': ['pdf']}}
+        response = self._get_manifest(ManifestFormat.curl, filters)
+        self.assertEqual(200, response.status_code, 'Unable to download manifest')
```
Comment (hannes-ucsc) [inline test/service/test_manifest.py:1205]:
```suggestion
        self.assertEqual(200, response.status_code)
```

Documenting or explaining the obvious actually makes code harder to read due to the increased verbosity.

File: test/service/test_manifest.py [hunk 7]
```diff
@@ -1195,6 +1195,28 @@ def test_full_metadata_missing_fields(self):
...
+            '--create-dirs',
+            '--compressed',
+            f'url="{config.service_endpoint()}/repository/files/5f9b45af-9a26-4b16-a785-7f2d1053dd7c"',
+            'output="SmartSeq2_RTPCR_protocol.pdf"',
+            f'url="{config.service_endpoint()}/repository/files/156c15a3-3406-45d3-a25e-27179baf0c59"',
+            'output="TissueDissociationProtocol.pdf"',
+            f'url="{config.service_endpoint()}/repository/files/0db87826-ea2d-422b-ba71-b15d0e4293ae"',
+            'output="SmartSeq2_sequencing_protocol.pdf"'
+        ]
+        self.assertEqual(set(lines), set(expected_lines))
```
Comment (hannes-ucsc) [inline test/service/test_manifest.py:1218]: This would fail to detect duplicate lines and broken orderings of lines. Let's discuss alternatives after stand-up.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/service/test_manifest.py [hunk 8]
```diff
@@ -1195,6 +1199,43 @@ def test_full_metadata_missing_fields(self):
...
+                'output="TissueDissociationProtocol.pdf"',
+                ''
+            ],
+            [
+                f'url="{config.service_endpoint()}/repository/files/5f9b45af-9a26-4b16-a785-7f2d1053dd7c"',
+                'output="SmartSeq2_RTPCR_protocol.pdf"',
+                ''
+            ]
+        ]
+        for expected, actual in zip_longest(expected_body, sorted(sliced(body, 3))):
```
Comment (hannes-ucsc) [inline test/service/test_manifest.py:1236]: Why do you compare each item of the outer list separately?

Comment (danielsotirhos) [reply test/service/test_manifest.py]: Ah, good point.

File: src/azul/service/manifest_service.py [hunk 9]
```diff
@@ -597,6 +600,34 @@ def create_file(self) -> Tuple[str, Optional[str]]:
...
+
+    def write_to(self, output: IO[str]) -> Optional[str]:
+        output.write('--create-dirs\n\n')
+        output.write('--compressed\n\n')
+        for hit in self._create_request().scan():
+            doc = self._hit_to_doc(hit)
+            file = one(doc['contents']['files'])
+            file_uuid = file['uuid']
+            file_name = file['name']
+            file_url = f'{config.service_endpoint()}/repository/files/{file_uuid}'
```
Comment (hannes-ucsc) [inline src/azul/service/manifest_service.py:625]: Should include version and must include catalog in URL.

File: test/service/test_manifest.py [hunk 10]
```diff
@@ -1195,6 +1196,36 @@ def test_full_metadata_missing_fields(self):
...
 
+    @manifest_test
+    def test_curl_format_manifest(self):
+        self.maxDiff = None
+        bundle_fqid = BundleFQID('f79257a7-dfc6-46d6-ae00-ba4b25313c10', '2018-09-14T133314.453337Z')
+        self._index_canned_bundle(bundle_fqid)
+        filters = {'fileFormat': {'is': ['pdf']}}
+        response = self._get_manifest(ManifestFormat.curl, filters)
+        self.assertEqual(200, response.status_code)
+        lines = response.content.decode('utf-8').splitlines()
```
Comment (hannes-ucsc) [inline test/service/test_manifest.py:1207]: 'utf-8' is the default

File: src/azul/service/manifest_service.py (hunk 9 above)
Comment (hannes-ucsc) [reply src/azul/service/manifest_service.py:625]: And use furl

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — COMMENTED** [review]
I'll push the fixes.

**Comment by hannes-ucsc** [comment]
For demo, simulate what the Data Browser would do with `/fetch/manifest/files` and show direct use via `/manifest/files`. Show idempotence after interrupted downloads.
